### ISSUE TITLE
感觉只用acc作为指标有点问题

### ISSUE THREAD

**Issue by DeepSai** [saiwaiyanyu/bi-lstm-crf-ner-tf2.0#9]
这种分类问题，使用precision 和 recall 作为 metrics 会比较好；假如训练数据中绝大多数都是O，那么acc一开始就会很高，不能准确反应模型的能力啊

**Comment by saiwaiyanyu** [comment]
是的，你提到这个是一个问题点。
在实际项目中，我们一般会以 完全标记正确的 Entity 计算 Recall，Precision， 评估会更客观。
这个模块暂时没有加，后面找时间补上去。

**Comment by ibmxiang** [comment]
paths = []
    accuracy = 0
    y_real = []


    for logit, text_len, labels in zip(logits, text_lens, labels_batch):
        viterbi_path, _ = tf_ad.text.viterbi_decode(logit[:text_len], model.transition_params)
        paths.append(viterbi_path)

        if type(labels) is np.ndarray:
            y_real.append(list(labels[:text_len]))
        else:
            y_real.append(list(labels.numpy()[:text_len]))

        correct_prediction = tf.equal(
            tf.convert_to_tensor(tf.keras.preprocessing.sequence.pad_sequences([viterbi_path], padding='post'),
                                 dtype=tf.int32),
            tf.convert_to_tensor(tf.keras.preprocessing.sequence.pad_sequences([labels[:text_len]], padding='post'),
                                 dtype=tf.int32)
        )

        accuracy = accuracy + tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        # print(tf.reduce_mean(tf.cast(correct_prediction, tf.float32)))
    accuracy = accuracy / len(paths)
    pre,recall = eval(paths,y_real)

**Comment by ibmxiang** [comment]
def eval(logits,labels_batch):
    entity = []
    pre_entity=[]
    for logit,label in zip(logits,labels_batch):
        enti =[]
        pre =[]
        for i in range(len(label)):
            if id2tag[label[i]]!='O':
                enti.append((id2tag[label[i]],i))
        for i in range(len(logit)):
            if id2tag[logit[i]]!="O":
                pre.append((id2tag[logit[i]],i))
        entity.append(enti)
        pre_entity.append(pre)
    # print(entity)
    # print(pre_entity)
    #预测的实体总数：
    pre_all = sum([len(i) for i in pre_entity])
    #真实的实体总数
    true_all = sum([len(i) for i in entity])
    #预测正确的：
    correct =0
    print(entity)
    print(pre_entity)
    for pre,true in zip(pre_entity,entity):
        inter = list(set(pre) & set(true))
        correct += len(inter)
    precission =correct/(pre_all+1)
    recall = (correct+1)/(true_all+1)
    # print(precission)
    # print(recall)
    return precission ,recall
